chalk-solve/src/clauses/builtin_traits/fn_family.rs

```diff
@@ -19,11 +20,116 @@ use chalk_ir::{
+                                    .cloned()
+                                    .map(|ty| ty.cast(interner)),
+                            ),
+                        }
+                        .intern(interner)
+                        .cast(interner);
+                        Substitution::from(interner, &[self_ty, arguments])
+                    };
+                    builder.push_fact(TraitRef {
+                        trait_id,
+                        substitution: substitution.clone(),
+                    });
+
+                    if let WellKnownTrait::FnOnce = well_known {
+                        let trait_datum = db.trait_datum(trait_id);
+                        let output_id = trait_datum.associated_ty_ids[0];
+                        let alias = AliasTy::Projection(ProjectionTy {
+                            associated_ty_id: output_id,
+                            substitution,
+                        });
+                        let return_type = inputs_and_output.return_type;
+                        builder.push_fact(Normalize {
+                            alias,
+                            ty: return_type,
+                        });
+                    }
+                });
+                Ok(())
+            }
+            TypeName::Closure(closure_id) => {
+                let closure_datum = db.closure_datum(closure_id, apply.substitution.clone());
+                let trait_matches = match well_known {
+                    WellKnownTrait::Fn => matches!(closure_datum.kind, ClosureKind::Fn),
+                    WellKnownTrait::FnMut => {
+                        matches!(closure_datum.kind, ClosureKind::FnMut | ClosureKind::Fn)
+                    }
+                    WellKnownTrait::FnOnce => matches!(
+                        closure_datum.kind,
+                        ClosureKind::FnOnce | ClosureKind::FnMut | ClosureKind::Fn
+                    ),
+                    _ => false,
+                };
+                if trait_matches {
+                    builder.push_binders(
+                        &closure_datum.inputs_and_output,
+                        |builder, inputs_and_output| {
+                            let substitution = Substitution::from(
+                                interner,
+                                Some(self_ty.cast(interner)).into_iter().chain(
+                                    inputs_and_output
+                                        .argument_types
+                                        .iter()
+                                        .cloned()
+                                        .map(|ty| ty.cast(interner)),
+                                ),
+                            );
+                            builder.push_fact(TraitRef {
+                                trait_id,
+                                substitution: substitution.clone(),
+                            });
+
+                            if let WellKnownTrait::FnOnce = well_known {
```

Comment:
...and especially this logic...feels like it is duplicated with the `FnDef` case, can we maybe extract a helper function?